### PR TITLE
fix/es6-destructuring-reword+new-lesson

### LINKED ISSUE #35574: Improvements to the challenge ES6 Use Destructuring Assignment to Assign Variables from Objects
Improvements to this challenge: https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects/

1. I suggest we remove the function from the challenge code. It just adds an extra layer of complexity with no added benefit. I suggest we just have an object and the destructuring.

```js
const AVG_TEMPERATURES = {
  today: 77.5,
  tomorrow: 79
};

const tempOfTomorrow = AVG_TEMPERATURES.tomorrow; // change this line to use destructuring instead
```

2. Instead of voxel and some random decimal numbers for the example code we should use something less abstract.

```js
var user = { name: 'John Doe', age: 34, password: 'qwerty' };

var name = user.name; // name = 'John Doe'
var age = user.age; // age = 34
var password = user.password; // password = 'qwerty'

const { name, age, password } = user; // name = 'John Doe', age = 34, password = 'qwerty'

const { name: userName, age: userAge, password: userPassword } = user; // userName = 'John Doe', userAge = 34, userPassword = 'qwerty'

const { objectProperty: variable } = object; // Same as const variable = object.objectProperty
```

3. The instructions should be clear. This is a bit tricky and I'm not 100% sure how to phrase it. If anyone has a suggestion please let me know.

> Use destructuring to assign the value of the tomorrow property from the object AVG_TEMPERATURES to a new variable named tempOfTomorrow. Look at the last code example given for help on syntax.

### PR DESCRIPTION
**Pull request by moT01**
- Added new lesson "Use Destructuring Assignment to Extract Values from Objects" to teach destructure without renaming variables.
- Rewrote "Use Destructuring Assignment to Assign Variables from Objects" to focus on destructure+renaming
- Rewrote "Use Destructuring Assignment to Assign Variables from Nested Objects" to align with the other lessons - and since the example there was abstract, I thought it could use it.

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #35574

### CODE REVIEW

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md [hunk 1]
```diff
@@ -6,85 +6,95 @@ challengeType: 1
 
 ## Description
 <section id='description'>
-We can similarly destructure <em>nested</em> objects into variables.
-Consider the following code:
+You can use the same principles from the previous two lessons to destructure values from nested objects.
+
+Using a similar object from previous examples:
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md]:
```suggestion
Using an object similar to previous examples:
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md [hunk 2]
```diff
@@ -6,85 +6,95 @@ challengeType: 1
...
 ```
 
-In the example above, the variable <code>startX</code> is assigned the value of <code>a.start.x</code>.
+Here's how to extract the values into variables with the same name as they have in the object:
+
+```js
+const { johnDoe: { age, email }} = user;
+```
+
+And here's how you can assign new variable names when destructuring a nested object like this:
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md]:
```suggestion
And here's how you can assign new variable names when destructuring a nested object:
```
I think this is a little better.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md [hunk 3]
```diff
@@ -6,85 +6,95 @@ challengeType: 1
...
+```js
+const { johnDoe: { age: userAge, email: userEmail }} = user;
+```
+
 </section>
 
 ## Instructions
 <section id='instructions'>
-Use destructuring assignment to obtain <code>max</code> of <code>forecast.tomorrow</code> and assign it to <code>maxOfTomorrow</code>.
+Replace the two assignments with an equivalent destructuring assignment. It should still assign the variables <code>lowToday</code> and <code>highToday</code> the values of <code>today.low</code> and <code>tomorrow.high</code> from the <code>LOCAL_FORECAST</code> object.
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md]:
```suggestion
Replace the two assignments with an equivalent destructuring assignment. It should still assign the variables <code>lowToday</code> and <code>highToday</code> the values of <code>today.low</code> and <code>today.high</code> from the <code>LOCAL_FORECAST</code> object.
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
This is a massive improvement on how it was done before, nice job!

Comments follow.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Thanks, @moT01, LGTM :+1:

**Comment by lasjorg** [comment]
Was asked to chime in. I think this looks great, so much better. :+1: 

I do have one question, do we want to use the word "field" when talking about an object property/key?

> You may read it as "get the field name and copy the value into userName" and so on.

This doesn't seem to match the terminology used in the curriculum, or am I wrong?


Other than that it LGTM, great work. @moT01

**Comment by moT01** [comment]
Thanks for the review @lasjorg - I changed it slightly

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md [hunk 4]
```diff
@@ -21,7 +21,7 @@ const { name: userName, age: userAge } = user;
 // userName = 'John Doe', userAge = 34
 ```
 
-You may read it as "get the field <code>name</code> and copy the value into <code>userName</code>" and so on.
+You may read it as "get the value of <code>user.name</code> and copy it into <code>userName</code>" and so on.
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md]:
```suggestion
You may read it as "get the value of <code>user.name</code> and assign it to a new variable named <code>userName</code>" and so on.
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@moT01 See my slight change in the verbiage.

**Review by RandellDawson — APPROVED** [review]

**Comment by lasjorg** [comment]
This might be splitting hairs, but I would prefer the use of `property` or `key` over `variable` when talking about objects.

**Comment by RandellDawson** [comment]
> This might be splitting hairs, but I would prefer the use of property or key over variable when talking about objects.

To which sentence are you referring?

**Comment by lasjorg** [comment]
All of them. Any mention of a property as a variable.

**Comment by RandellDawson** [comment]
@lasjorg I see what you mean now.  Feel free to request how you would word them differently.  I would support it.

**Comment by lasjorg** [comment]
Not really sure how to format this.

https://github.com/freeCodeCamp/freeCodeCamp/blob/12527818dd4e63c6cb4bfbb3c3ed910416224311/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md

> - Here's how to extract the values into variables with the same name as they have in the object:
> - Here's how to extract the values into properties with the same name as they have in the object:


> - And here's how you can assign new variable names when destructuring a nested object:
> - And here's how you can assign new property names when destructuring a nested object:


> - Replace the two assignments with an equivalent destructuring assignment. It should still assign the variables lowToday and highToday the values of today.low and today.high from the LOCAL_FORECAST object.
> - Replace the two assignments with an equivalent destructuring assignment. It should still assign the properties lowToday and highToday the values of today.low and today.high from the LOCAL_FORECAST object.



https://github.com/freeCodeCamp/freeCodeCamp/blob/f743d5c0738668ba6c5372d741979235601fc94f/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md

> - Destructuring allows you to assign a new variable name when extracting values. You can do this by putting the new name after a colon when assigning the value.
> - Destructuring allows you to assign to a new property when extracting values. You can do this by putting the new property name after a colon when assigning the value.


> - Here's how you can give new variable names in the assignment:
> - Here's how you can give new property names in the assignment:


> - You may read it as "get the value of <code>user.name</code> and assign it to a new variable named <code>userName</code>" and so on.
> - You may read it as, get the value from the property `name` and assign it to a new property named `userName`.


> - Replace the two assignments with an equivalent destructuring assignment. It should still assign the variables highToday and highTomorrow the values of today and tomorrow from the HIGH_TEMPERATURES object.
> - Replace the two assignments with an equivalent destructuring assignment. It should still assign the properties highToday and highTomorrow the values of today and tomorrow from the HIGH_TEMPERATURES object.

**Comment by RandellDawson** [comment]
@lasjorg I was thinking you were going in a different direction.

The first one (seen below):
>`Here's how to extract the values into properties with the same name as they have in the object:

You are not extracting values into properties.  You are assigning the property values to variables. 
I was thinking something more like:
>Here's how to extract the values of object properties and assign them to variables with the same name as the properties of the object:

The same kind of verbiage would apply to the other sentences (see an example below):

> And here's how you can assign an object properties' values to variables with different names than the  proprerty names of the nested object:

**Comment by lasjorg** [comment]
@RandellDawson  Your right I don't know what the hell I was thinking. That was a massive brain fart. I just started to replace all the variables with properties, I completely forgot what the challenges were even doing. My only excuse is it's late and I clearly shouldn't be anywhere near code right now. Sorry, that was embarrassing. :disappointed:

**Comment by moT01** [comment]
What's the problem here? I'm not sure I see it.

**Comment by scissorsneedfoodtoo** [comment]
Thank you @moT01 for taking the time to add this lesson and improve the existing ones. These are a huge improvement.

I'm in favor of changing the wording in `/use-destructuring-assignment-to-assign-variables-from-nested-objects.english.md` slightly as @RandellDawson suggested. Perhaps something just slightly shorter like "Here's how to extract the values of object properties and assign them to variables with the same name:", and the following sentence to "And here's how you can assign an object properties' values to variables with different names:"

**Review by scissorsneedfoodtoo — APPROVED** [review]

**Comment by scissorsneedfoodtoo** [comment]
Thank you @moT01 -- everything LTGM :1st_place_medal:

**Pull request merged**